src/Compilers/VisualBasic/Portable/Semantics/Conversions.vb

```diff
@@ -916,13 +916,16 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
             userDefinedConversionsMightStillBeApplicable = False
             Dim conv As ConversionKind
 
-            ' Using <symbol>.IsConstant() for field accesses can result in an infinite loop.
+            ' Using source.IsConstant for field accesses can result in an infinite loop.
             ' To detect such a loop pass the already visited constants from the binder.
+            ' The check for source.IsConstant is still necessary because the node might still 
+            ' be considered as a non-constant in some error conditions (a reference before declaration,
+            ' for example).
             Dim sourceIsConstant As Boolean = False
             If source.Kind = BoundKind.FieldAccess Then
-                sourceIsConstant = DirectCast(source, BoundFieldAccess).FieldSymbol.GetConstantValue(binder.ConstantFieldsInProgress) IsNot Nothing
+                sourceIsConstant = DirectCast(source, BoundFieldAccess).FieldSymbol.GetConstantValue(binder.ConstantFieldsInProgress) IsNot Nothing AndAlso source.IsConstant
```

Comment:
I don't believe we can have the same situation for fields right now and making defensive change because it can only make things better, not worse.